ext/storage_sources/s3_store/s3_store.cpp

```diff
@@ -0,0 +1,253 @@
+                if (bucket_objects.size() >= 1) {
+                    for (Aws::S3Crt::Model::Object &object : bucket_objects) {
+                        std::cout << "  * " << object.GetKey() << std::endl;
+                    }
+                } else {
+                    std::cout << "No objects in bucket." << std::endl;
+                }
+                std::cout << std::endl;
+            }
+
+            /* Put object. */
+            fs->conn->put_object(first_bucket, "WiredTiger.turtle", "WiredTiger.turtle");
+
+            /* List objects again. */
+            if (fs->conn->list_objects(first_bucket, bucket_objects)) {
+                std::cout << "Objects in bucket '" << first_bucket << "':" << std::endl;
+                if (bucket_objects.size() >= 1) {
+                    for (Aws::S3Crt::Model::Object &object : bucket_objects) {
+                        std::cout << "  * " << object.GetKey() << std::endl;
+                    }
+                } else {
+                    std::cout << "No objects in bucket." << std::endl;
+                }
+                std::cout << std::endl;
+            }
+
+            /* Delete object. */
+            fs->conn->delete_object(first_bucket, "WiredTiger.turtle");
+
+            /* List objects again. */
+            if (fs->conn->list_objects(first_bucket, bucket_objects)) {
+                std::cout << "Objects in bucket '" << first_bucket << "':" << std::endl;
+                if (bucket_objects.size() >= 1) {
+                    for (Aws::S3Crt::Model::Object &object : bucket_objects) {
+                        std::cout << "  * " << object.GetKey() << std::endl;
+                    }
+                } else {
+                    std::cout << "No objects in bucket." << std::endl;
+                }
+                std::cout << std::endl;
+            }
+        } else {
+            std::cout << "No buckets in AWS account." << std::endl;
+        }
+    }
+
+    *file_systemp = &fs->file_system;
+   return 0;
+}
+
+/*
+ * s3_fs_terminate --
+ *      Discard any resources on termination of the file system.
+ */
+static int
+s3_fs_terminate(WT_FILE_SYSTEM *file_system, WT_SESSION *session)
+{
+    S3_FILE_SYSTEM *s3_fs;
+
+    UNUSED(session); /* unused */
+
+    s3_fs = (S3_FILE_SYSTEM *)file_system;
+    delete(s3_fs->conn);
+    free(s3_fs);
+
+    return (0);
+}
+
+/*
+ * s3_add_reference --
+ *     Add a reference to the storage source so we can reference count to know when to really
+ *     terminate.
+ */
+static int
+s3_add_reference(WT_STORAGE_SOURCE *storage_source)
+{
+    UNUSED(storage_source);
+    return (0);
+}
+
+/*
+ * s3_terminate --
+ *     Discard any resources on termination.
+ */
+static int
+s3_terminate(WT_STORAGE_SOURCE *storage, WT_SESSION *session)
+{
+    /* NEED A WAY to pass the options to the ShutdownAPI call */
+    S3_STORAGE *s3;
+    s3 = (S3_STORAGE *)storage;
+
+    Aws::ShutdownAPI(options);
+
+    free (s3);
```

Comment:
It's weird that we pass in something and it is freed in the function. Should the caller of this function manage the lifecycle of storage?

Reply:
This function will eventually do more. Have a look at the following local store equivalent to get an idea:
https://github.com/wiredtiger/wiredtiger/blob/develop/ext/storage_sources/local_store/local_store.c#L1120

For now, this function is blindly freeing the storage source.